docs: refactor live-patching database docs. (#558)
canonical/launchpad-manual: docs/developer/explanation/database/database-patching.rst

.. meta::
   :description: Understand database patching techniques for deploying 
      schema changes on Launchpad's PostgreSQL database.

.. _database-patching:

Database patching
=================

This page describes the schema changes that we can deploy on Launchpad's

> replicated PostgreSQL database, and the two mechanisms we use to apply them:

Reviewer: ```suggestion
This page describes the schema changes that can be deployed on Launchpad's
replicated PostgreSQL database, and the two mechanisms used to apply them:
```
3rd person

- **Hot schema changes** deploy changes without causing any downtime.
- **Cold schema changes** deploy changes with minimal downtime (a few seconds
  per change).

In both cases, we deploy the changes without altering the running appserver
code, so each change must be compatible with the existing live code. This means
that, before landing a database patch, we may need to land code first so that
everything runs both with and without the patch applied. For the end-to-end

> workflow, see :ref:`database-schema-changes-process`.

Reviewer: ```suggestion
In both cases, the changes are deployed without altering the running appserver
code, so each change must be compatible with the existing live code. This means
that, before landing a database patch, you may need to land code first so that
everything runs both with and without the patch applied. For the end-to-end
workflow, see :ref:`database-schema-changes-process`.
```

> We land cold schema changes on ``db-devel`` and hot ones on ``devel``. All

Reviewer: ```suggestion
Cold schema changes  are landed on ``db-devel`` and hot ones on ``devel``. All
```

> schema changes must land without appserver code changes, although test changes

Reviewer: Is there some contradiction here? We previously mention the possibility of landing code changes before landing db patches. This now says schema changes must land without appserver code changes. What am I missing?

Author: I think in most cases - we land the patch first and then follow up with the code changes. I have not yet ran into cases where we had to land the code changes first.

Author: That said I think the crux of the previous statement is to ensure the existing code is compatible with any DB changes you plan.

may be included when trigger or database function changes are being landed. For
more on developing against the ``db-devel`` branch, see
:ref:`working-with-db-devel`.

.. note::

   These instructions only apply to tables under native PostgreSQL streaming
   replication. ``lp_*`` tables are replicated to SSO via Slony, so they
   require extra special care. See :ref:`postgresql-and-launchpad` for more on
   how Launchpad uses replication.

Hot patches
-----------

We can apply some database patches live. We currently apply them by crafting an
appropriate wrapper script and then running it directly on the master database

> node, from where it replicates to the slaves.

Reviewer: ```suggestion
Some database patches can be applied live. These are currently applied by crafting an
appropriate wrapper script and then running it directly on the master database
node, from where it replicates to the slaves.
```

The ``(major, minor, patch)`` revision tuple must be manually inserted into the

> ``LaunchpadDatabaseRevision`` table on the master. Failing to do this aborts the

Reviewer: Who is responsible for manually inserting the revision tuple into the table?

Author: Launchpad automation does that as part of fastdowntime.

next full upgrade because the system attempts to apply a patch that has already
been applied. This situation is normally caught by the full staging update.

We intend to automate this in the future, to at least handle the happy path and

> avoid accidentally running a hot patch during a fastdowntime deployment.

Reviewer: ```suggestion
This should be automated in the future, to at least handle the happy path and
avoid accidentally running a hot patch during a fastdowntime deployment.
```

Index creation
~~~~~~~~~~~~~~

Index creation for all but the smallest tables should be done in a hot patch.

The patch must use ``CREATE INDEX CONCURRENTLY``, which runs three separate fast
transactions to create, populate, and activate the index. This adds some
contention, but it is minimal.

> Don't start index creation while there are long-running transactions in

Author: ```suggestion
Don't start index creation while there are long-running transactions (confirm from the DBA) in
```

progress: the index will not complete until those transactions have completed
because it has to wait for the existing read transactions on the table to finish
before it can be activated.

``CREATE INDEX CONCURRENTLY`` cannot run inside a transaction. If the index
cannot be built for some reason, such as trying to build a ``UNIQUE`` index on
non-unique data, manual repair is needed to roll back the changes.

.. warning::

   Deleting an index requires a fastdowntime deployment.

Run index creation from a ``screen(1)`` session or similar to avoid network
dropouts. For ``CREATE INDEX CONCURRENTLY``, an early disconnection results in
an invalid index rather than a rollback.

Once the index is built on the master, it replicates automatically to the
slaves. After creating the indexes, confirm that there are no invalid indexes 
on the affected tables.

Views
~~~~~

> Run the ``CREATE OR REPLACE VIEW`` statements on the master.

Reviewer: What's the objective of running this? i.e., "Run the `CREATE OR REPLACE VIEW` statements on the master to...{create a view/replace a view/something else..."

Author: It basically mentions that if you have to create a view while DB patching, run the "create view" command on master. 

Just to be epxlicit that its fine to run this command directly.

   replication. ``lp_*`` tables are replicated to SSO via Slony, so they
   require extra special care. See :ref:`postgresql-and-launchpad` for more on
   how Launchpad uses replication.

Hot patches
-----------

We can apply some database patches live. We currently apply them by crafting an
appropriate wrapper script and then running it directly on the master database
node, from where it replicates to the slaves.

The ``(major, minor, patch)`` revision tuple must be manually inserted into the
``LaunchpadDatabaseRevision`` table on the master. Failing to do this aborts the
next full upgrade, because the system attempts to apply a patch that has already
been applied. This situation is normally caught by the full staging update.

We intend to automate this in the future, to at least handle the happy path and
avoid accidentally running a hot patch during a fastdowntime deployment.

Index creation
~~~~~~~~~~~~~~

Index creation for all but the smallest tables should be done in a hot patch.

The patch must use ``CREATE INDEX CONCURRENTLY``, which runs three separate fast
transactions to create, populate, and activate the index. This adds some

> contention, but it is minimal.

Reviewer: What does contention mean in this context?

Author: Contention to hold locks on the rows.

> Don't start index creation while there are long-running transactions in

Reviewer 2: is this still true with `CONCURRENTLY`?

Reviewer: Still on this line, how would someone know if there are long-running transations?

Author: They will have to ask the DBA (someone who can queries on prod DB directly)

Postgres surfaces information on whether there are any long running transactions.

Reviewer 2: alternatively, we are allowed to run preflight checks, that include the long running transactions check

Reviewer: If there is no obvious way for the user to know and if they have to ask the DB administrator to check, please add some direction there, e.g., ` (confirm from the DBA)`

Author: added.

Run the ``CREATE OR REPLACE VIEW`` statements on the master.

Stored procedures
~~~~~~~~~~~~~~~~~

Stored procedures can be changed, as long as the signature is still consistent
with any triggers. Table trigger definitions must be done as cold patches.

Cold patches
------------

Cold patches must be applied with no other system activity, to prevent locks
from holding the patch up. We have automation around this to apply them as
rapidly as possible.

Patch process overview
~~~~~~~~~~~~~~~~~~~~~~

1. Update the ``wildcherry`` Launchpad source tree to the revision to run. Make
   sure that the revision will include only one new patch. We apply one patch
   at a time to reduce the debugging burden when something goes wrong after
   applying it, and run ``make clean build``.

2. Run ``full-update.py --dry-run`` and check that only the expected database
   patch will be applied.

3. Run ``full-update.py``.

Table creation
~~~~~~~~~~~~~~

Creating a table can sometimes be done hot, but not reliably, because foreign
key relations need exclusive table locks to activate. Do it as a cold patch.

.. warning::

   Don't attempt to create a table without the DBA available. There are still

>    edge cases where we cannot do this live, such as a table with foreign key

Reviewer 2: if it is not live, why the page is called live patching?

Reviewer 2: also, who's the DBA?

Author: will update to database-patching. There's no official DBA, Clint takes care of things for now. I believe we can just leave it as it is and not mention specific people who has access.

   constraints to a busy table, which requires locks on the busy table that
   will not be granted quickly enough.

Table renaming
~~~~~~~~~~~~~~

Because code is deployed separately from database updates, we need to maintain a
fully updateable working alias from the old table name to the new one. See
:ref:`rename-database-table` for an example database patch.

Adding columns
~~~~~~~~~~~~~~

If the table is large enough that applying the patch raises performance
concerns, split the work into the following steps. All steps aside from the
first are optional, depending on the desired result.

1. Create the column without indexes and without setting ``NOT NULL`` or a
   default value.
2. Set the default value.
3. Add the index via a hot patch.
4. Run a garbo job (or similar) to fill in the values.
5. Set the column ``NOT NULL``.

.. note::

   We still need to work out how we test that things work both with and without

>    the database patch applied.

Reviewer: I don't know how I feel about this admonition. We stated earlier that code changes should work with or without the database patch. Stating here that we don't have a way of testing this reveals a significant gap.

Author: I think this comment is more on the pre-prod testing side. afaik our testing pipeline applies all the patch and then tests the code - so this comment still seems to be relevant.

Author: You are right, that's a gap on the testing side.
